enhancements/network/ovn-kubernetes-localnet-api.md

```diff
@@ -0,0 +1,540 @@
+    // Layer2 topology creates one logical switch shared by all nodes.
+    // Localnet topology attach to the overlay local network. Enables egress to the provider's physical network. 
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    // +unionDiscriminator
+    Topology NetworkTopology `json:"topology"`
+    
+    ...
+    
+    // Localnet is the Localnet topology configuration.
+    // +optional
+    Localnet *LocalnetConfig `json:"localnet,omitempty"`
+}
+```
+
+The CUDN spec should have additional validation rule for `spec.network.topology` field:
+```go
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+    ...
+    // +required
+    Network NetworkSpec `json:"network"`
+}
+```
+
+As of today the CUDN has validation to ensure `spec.network` is immutable:
+```go
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Network spec is immutable"
+```
+This validation should be changed in a way it allows mutating localnet topology configurations, 
+at least MTU, VLAN, physicalNetworkName and excludeSubnets. 
+The proposed change is to invert the immutable-spec validation on CUDN `spec.network` as follows:
+1. Remove mutation restrictions on `spec.network`.
+2. Add CEL rule validations to restrict mutations of:
+   `spec.network.topology`, `spec.network.layer2`, and `spec.network.layer3`.
+3. Add CEL rule validations to restrict mutation of localnet topology settings: 
+   `spec.network.localnet.role`, `spec.network.localnet.subnetes` and `spec.network.topology.ipam`.
+
+Localnet topology configuration type 
+```go
+// +kubebuilder:validation:XValidation:rule="has(self.ipam) && has(self.ipam.mode) && self.ipam.mode != 'Enabled' || has(self.subnets)", message="Subnets is required with ipam.mode is Enabled or unset"
+// +kubebuilder:validation:XValidation:rule="!has(self.ipam) || !has(self.ipam.mode) || self.ipam.mode != 'Disabled' || !has(self.subnets)", message="Subnets must be unset when ipam.mode is Disabled"
+// +kubebuilder:validation:XValidation:rule="!has(self.excludeSubnets) || has(self.excludeSubnets) && !has(self.subnets)", message="excludeSubnets must be unset when subents is unset"
+// +kubebuilder:validation:XValidation:rule="!has(self.subnets) || !has(self.mtu) || !self.subnets.exists_one(i, isCIDR(i) && cidr(i).ip().family() == 6) || self.mtu >= 1280", message="MTU should be greater than or equal to 1280 when IPv6 subent is used"
+type LocalnetConfig struct {
+    // role describes the network role in the pod, required.
+    // Whether the pod interface will act as primary or secondary. 
+    // For Localnet topology only `Secondary` is allowed.
+    // Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
+    // +required
+	// +kubebuilder:validation:Enum=Secondary
```

Comment:
If `NetworkRole` already specifies an enum validation tag this won't work. To use an Enum tag this needs to be its own type alias to a string or a string. The only other path to restrict this further in this case would be a CEL expression.

Reply:
using a CEL expression you would end up with a potentially misleading CRD schema though because `Primary` would show up as an allowed value in the schema but in reality it would be blocked by your XValidation

Reply:
I think we can avoid it by moving the enum valudation for NetworkRole defintion to the toplogies defintions (where NetworkRole used), as follows:
1. Remove NetworkRole enum validation
2. On Layer2Config definition, add  enum validation to `role` field allowing `Primary` or `Secondary`.
3. On Layer3Config definition, add  enum validation to `role` field allowing `Primary` or `Secondary`.
4. Finally, in the proposed  LocalnetConfig definition, add  enum validation to `role` field allowing `Secondary` only.

@everettraven WDYT?

Reply:
Added text about required changes to topology role validations https://github.com/openshift/enhancements/compare/6fc5b82ece4e9bdaaedf6d7208f56681dc40bcdc..7caa6a8f94524aa3218f9ea2ef845bcc377e4893#diff-65ff4f11b53425ec17e6fbd228843c58470c309f5031c38c345010775bb378d9R508-R519